Add macros to create and run stored procedure on Snowpark

zloff opened the pull request:
## Description & motivation

This change to run the python code on Snowpark is related to https://snplow.atlassian.net/browse/PE-4887 . 

There isn't much detail around specifics of how this would integrate with the package, so I created two on-run-end hooks that create and run the stored procedure if the user is running on snowflake and they wish to run the python with snowpark (I created a boolean var for this purpose). 

The python code I put as a large block in the create stored procedure macro - both the main_snowplow_snowflake.py and fractribution.py file contents are in there. I just had to adjust the code regarding session - the session must be passed in rather than created in the script. 

I'm all ears if you have some suggestions as to how to do this better!


## Checklist
- [x ] I have verified that these changes work locally
- [ ] I have updated the README.md (if applicable)
- [ ] I have added tests & descriptions to my models (and macros if applicable)
- [ ] I have raised a [documentation](https://github.com/snowplow/documentation) PR if applicable (Link here if required)

<!-- 
## Release Only Checklist
- [ ] I have updated the version number in all relevant places
- [ ] I have updated the CHANGELOG.md 
-->

zloff commented on macros/snowflake_snowpark/create_stored_procedure.sql:
```diff
@@ -0,0 +1,514 @@
+
+{% macro snowflake__create_report_table_proc() %}
+{% set stored_proc %}
+create or replace procedure {{target.schema}}_derived.create_report_table(attribution_model STRING, conversion_window_start_date STRING, conversion_window_end_date STRING)
```
How do I get this to create the procedure in `_derived` without hardcoding it...?

rlh1994 replied:
I'd have to double check this, but I _think_ the `target.schema` is relative to the model that called the macro itself, however if it's called via a hook then I think it's just the default profile target so you'd have to hardcode it...

zloff replied:
Okay so i've figured out that I should actually be using `{{schema}}` not `{{target.schema}}` as the former compiles to the custom schema without any hardcoding. And it seems to work whether it's in the on-run-end in dbt_project.yml or a post_hook in a model config.

rlh1994 commented on dbt_project.yml:
```diff
@@ -42,8 +42,15 @@ vars:
+  
 on-run-start: "{{ create_udfs() }}"
+on-run-end: 
+  - "{% if target.type == 'snowflake' and var('snowplow__run_python_script_in_snowpark') %}{{ create_report_table_proc() }}{% endif %}"
+  - "{% if target.type == 'snowflake' and var('snowplow__run_python_script_in_snowpark') %}{{ run_stored_procedure(var('snowplow__attribution_model_for_snowpark'), var('snowplow__conversion_window_start_date') , var('snowplow__conversion_window_end_date') ) }}{% endif %}"
```
This makes me sad, could we not just always trigger the macros and they do nothing on the non-snowflake warehouses?

zloff replied:
Yeah can do! I guess I just figured this makes it a little more obvious that the hooks only do something on Snowflake, whereas on the other warehouses you'd have to go check the macro itself to see that nothing is actually being done.

Also I figured that we should give the Snowflake user the option of not running this on snowpark in case they wanted to do it themselves. But i'm not sure how many users would actually want to run it themselves if there is this easy option anyway?

zloff replied:
One thing to consider is that these hooks will run even if the models fail, which I don't like. It might be another vote for a hook in the config of a model instead, but I agree with you below, it is a little more 'hidden' to do it that way.

rlh1994 replied:
> Also I figured that we should give the Snowflake user the option of not running this on snowpark in case they wanted to do it themselves.

Yeah I think keep that variable as well!

rlh1994 commented:
To fix the existing tests you'll need to make sure the schema is created if it doesn't exist before you add the macro to it. It's currently triggering the `on-run-end` on the `seed` call, before it has created the `derived` schema in the normal `run`. I'm not sure if there's a way to only run a hook on a specific type of call, as we really don't need this running on seeds. Maybe you can fix this with an `if execute` block around the contents of the macros?

If we are going to add this to the package we should also add integration tests to snowflake for the final output table(s) that get created. Let me know if you need a hand doing this.

In terms of alternative approaches, I think this makes sense and if it works then great! It would be nice if we can tidy and sort the schema location stuff - it _might_ work as a post-hook on a specific model and use that models full schema name, but I have no real idea because as always dbt doesn't tell you these things, and then we're basically hiding the creation of this in another model...

zloff commented:
> 

Ah ok, I will add an if execute block. Thanks!

re: integration tests, I haven't done any before but will have a go.

I'm still unsure as to whether this should be an on-run-end or a post_hook, there seems to be advantages/disadvantages to both - the largest is that if associated with a model, it won't run if a downstream model fails, whereas an on-run-end hook will run regardless. But it is less clear what's going on. Jethro also suggested just having an inactive model that calls the macros that the user has to specifically run at the end. That could work too.

For now, i've added them to the path_summary model since i've been playing around, but happy to move them wherever is best.

rlh1994 commented on models/snowplow_fractribution_path_summary.sql:
```diff
@@ -1,6 +1,8 @@
   config(
-    sql_header=snowplow_utils.set_query_tag(var('snowplow__query_tag', 'snowplow_dbt'))
+    sql_header=snowplow_utils.set_query_tag(var('snowplow__query_tag', 'snowplow_dbt')),
+    post_hook= ["{% if target.type == 'snowflake' and var('snowplow__run_python_script_in_snowpark') %}{{ create_report_table_proc() }}{% endif %}",
+    "{% if target.type == 'snowflake' and var('snowplow__run_python_script_in_snowpark') %}{{ run_stored_procedure(var('snowplow__attribution_model_for_snowpark'), var('snowplow__conversion_window_start_date') , var('snowplow__conversion_window_end_date') ) }}{% endif %}"]
```
Can this be done in a single if block, maybe with a semi colon between them?

rlh1994 commented:
> re: integration tests, I haven't done any before but will have a go.

Great, you should just need to add a dataset for the expected output, based on the test data, and then add the test condition, hopefully you can just copy from what we already have and adapt it but let me know if you need a hand or anything is unclear!

> For now, i've added them to the path_summary model since i've been playing around, but happy to move them wherever is best.

I think that model makes sense, assuming there is no dependency on the channel count/spends, as otherwise there is no certainty those tables will have been built before. If it does depend on those then you'll need to force a dependency https://docs.getdbt.com/reference/dbt-jinja-functions/ref#forcing-dependencies

![image](https://github.com/snowplow/dbt-snowplow-fractribution/assets/8260415/d22d8a35-8f64-4d9f-877a-8e05009d3a53)

zloff commented:
> > re: integration tests, I haven't done any before but will have a go.
> 
> Great, you should just need to add a dataset for the expected output, based on the test data, and then add the test condition, hopefully you can just copy from what we already have and adapt it but let me know if you need a hand or anything is unclear!

I have added integration tests, but have had to add some workarounds. For one - the tables aren't created by dbt and therefore have no corresponding model file, so I added references to them using source() instead of ref(). Secondly, in the models in the `actual` folder, I had to add a dummy select statement if the target wasn't snowflake so that it didn't try to find the tables. 
The tests can only be run when `target.type= snowflake` as the tables won't exist in other target types as soon as the dbt package has run (as the python script is run manually). So I added an `enabled` config on the new tests. 

Let me know if any of this is confusing or if you have suggestions for doing this better!

I'll address the other comments tomorrow morning :)

rlh1994 commented:
That all looks good to me! A worrying chunk of our tests are "we have to do it this way to get it working". Perhaps a dumb question, but where are the sources defined?

zloff commented:
> 

Ah, I added them to the bottom of `actual_vs_expected.yml`

zloff commented:
> > re: integration tests, I haven't done any before but will have a go.
> 
> Great, you should just need to add a dataset for the expected output, based on the test data, and then add the test condition, hopefully you can just copy from what we already have and adapt it but let me know if you need a hand or anything is unclear!
> 
> > For now, i've added them to the path_summary model since i've been playing around, but happy to move them wherever is best.
> 
> I think that model makes sense, assuming there is no dependency on the channel count/spends, as otherwise there is no certainty those tables will have been built before. If it does depend on those then you'll need to force a dependency https://docs.getdbt.com/reference/dbt-jinja-functions/ref#forcing-dependencies

Okay I've changed the strategy because I didn't want to add a `--depends_on` to the path_summary model when it actually doesn't depend on channel_spend. Instead of calling the macros as a post-hook in the path_summary model, I now have an extra model definition (`snowplow_fractribution_call_snowpark_macros`) that calls the macros. I originally had this as ephemeral but when the models were run, the cli output didn't mention that it was being run, so it just silently created and called the stored procedure. So i materialised as a table, but then had to actually output something for a table - so I now output the datetime of the run....

I swear these integration tests will be the death of me. In order for them to work properly on each warehouse, I added `snowplow__run_python_script_in_snowpark: "{{ target.type == 'snowflake' | as_bool() }}"` to the dbt_project.yml. BUT this evaluates to 'True' which the model config in `call_snowpark_macros` doesn't recognise as bool, so I added an ugly conditional statement. Please suggest something better haha

zloff commented:
I'm not sure what happened with databricks, I cancelled the pr_test after almost 2 hours...

rlh1994 commented:
Databricks is having some issues this week don't worry about it 😅

rlh1994 commented on models/snowplow_fractribution_call_snowpark_macros.sql:
Wonder if it's worth adding a log into this to tell the user what tables are being created where when this model runs each time?

zloff replied:
oh yeah, can do. So the output table of this model would contain something like the table name, dataset and last_created timestamp for the three tables that are created in snowpark?

zloff replied:
Or by log do you mean something that is printed to the console?

rlh1994 replied:
I meant log to the console, but having them in the table is also a good idea!

rlh1994 commented:
It would make the tests longer because you'd basically have to run the fract part and the tests twice, but I wonder if it's worth finding a way to test a different attribution method as well, to ensure that variable is working as intended. 

You might be able to "hack" it by adding a column to your expected seed data with the attribution type, then filter in the `_stg` query based on the variable value and not select that column. Hopefully that makes sense if not just let me know.

rlh1994 commented on LICENSE:
Just want to check we can actually do this @miike? Given we based some of this off Google and are building on-top of dbt, are we allowed to use this license?

rlh1994 commented:
Last few things left (just grouping previous comments):
 - [x] Update tests to test an additional attribution method as well
 - [x] Add console log and columns to the model that generates additional tables to be explicit
 - [x] Test that the docs generate and serve locally 
 - [ ] Confirm we can use the new license in this case

Docs PR already raised here for when we release: https://github.com/snowplow/documentation/pull/460

zloff commented:
> Last few things left (just grouping previous comments):
> 
> * [ ]  Update tests to test an additional attribution method as well
> * [ ]  Add console log and columns to the model that generates additional tables to be explicit
> * [ ]  Test that the docs generate and serve locally
> * [ ]  Confirm we can use the new license in this case


> 
> Docs PR already raised here for when we release: [snowplow/documentation#460](https://github.com/snowplow/documentation/pull/460)

Thanks Ryan! I've updated the tests to use a second attribution method, added the console log/update to the output table. Docs seem to generate locally, except the console log message is printed when building the docs:

<img width="617" alt="Screen Shot 2023-05-29 at 3 48 59 pm" src="https://github.com/snowplow/dbt-snowplow-fractribution/assets/61371935/a06fe105-2a81-48b4-b365-902a84a9d1bb">
I added an if execute block (because it was being printed on compilation) but that didn't help with this issue. Any suggestions?

rlh1994 commented:
> Any suggestions?

I don't think there is an obvious way around it, we have the same issue with the date range log as you see, and it's not really that much of an issue anyway. ~**Maybe** could get around it by:~
~- Adding the log to the run macro, rather than directly in the model~
~- Maybe adding it as a post hook, not sure if these get run on docs generation?~

rlh1994 reviewed (changes requested):
I'm not sure _What_ but this returns less rows for the final report table than when I run it using the existing method on snowflake, in particular all rows with 0 conversions/revenue/roas are not available in the snowpark output table but are in the normal method. They are also missing from the channel attribution table. They seem to match the path summary with channels though.

zloff commented:
> I'm not sure _What_ but this returns less rows for the final report table than when I run it using the existing method on snowflake, in particular all rows with 0 conversions/revenue/roas are not available in the snowpark output table but are in the normal method. They are also missing from the channel attribution table. They seem to match the path summary with channels though.

Hmm strange. I spent some time looking into this and couldn't specifically replicate. When I ran using snowpark vs docker using the shapley method, both returned a table with three rows. But then when I tested with linear, both methods returned 6 rows, with zeroes in the extra rows. Are you sure that the attribution method was the same both times you ran it? Nevertheless it does bear looking into as to the different output between methods. 
I also got errors when i tried to run last_touch and first_touch - do you also get DecimalType errors for those? I haven't had a chance to look into them further

rlh1994 commented:
Ah right, I was testing on `last_touch` and had fixed 2 bugs for that (https://github.com/snowplow/dbt-snowplow-fractribution/pull/15/files#diff-8952aee9545be062beea9532406debe7691446504c30e9be776ff045103fbfe9) so that it outputs.

I'll double check what I had today, and also refresh myself on what the different methods do and see if that's why they give different numbers of records...

rlh1994 commented:
I've re-run it this morning, running with `last_touch` it seems that the snowpark method returns just the 3 rows, the standard method returns 6. You're right that shapley seems to return 3 for both methods.

I assume it's related to this line in the shapely def (which I assume is intended?):
https://github.com/snowplow/dbt-snowplow-fractribution/blob/ed04da85fcd635520dc46569d236671e4dfc07cd/utils/fractribution.py#L134-L140 
It may be that technically these should be removed from other paths as well so it's consistent - but maybe let's leave that for another PR

Unfortunately our integration tests give the exact same output for every model type so it's a little hard to debug what is actually happening (i.e. is it not choosing the right model, or is it just on a different python version that for some reason is throwing away these records)...

Edit:
I've tried to mess around with the source data to try and generate different outputs from each model but to be honest I can't quite follow everything through to work out exactly what to change. I'd recommend do that first - so using the python script method gives different outputs (at least for one of the rows) between two methods, then that will make it much easier to compare what is happening with the snowpark one.

rlh1994 commented:
Playing around with this more it gets even weirder, it seems that BigQuery does this in a different way where Shapley has 6 rows but linear for example only has 3. But if I run the query that generates the table it returns 6 rows. I feel like it can only be to do with the connection packages but I don't know what would make this happen on some models and not others....

I think there's a decision to make that should make this easier to sort, should the final report table contain channels with 0 conversions? If so, we need to change the python code to allow them and potentially work out what is happening to remove them in the output, if not we could be consistent by adding a where clause to the SQL that creates the report table and remove them regardless of model.

What do you think is the right thing to do? My opinion would be that we should include them as if there was a spend on that channel it's still important to report this.

zloff commented:
> Playing around with this more it gets even weirder, it seems that BigQuery does this in a different way where Shapley has 6 rows but linear for example only has 3. But if I run the query that generates the table it returns 6 rows. I feel like it can only be to do with the connection packages but I don't know what would make this happen on some models and not others....
> 
> I think there's a decision to make that should make this easier to sort, should the final report table contain channels with 0 conversions? If so, we need to change the python code to allow them and potentially work out what is happening to remove them in the output, if not we could be consistent by adding a where clause to the SQL that creates the report table and remove them regardless of model.
> 
> What do you think is the right thing to do? My opinion would be that we should include them as if there was a spend on that channel it's still important to report this.

Yeah I agree - spend on a channel that's generating no revenue is important to know. 
I had a bit of a look at this today too but didn't really figure anything out - i've been swamped by GA4 migration work! I added the bug fixes for last and first touch that you linked to above to the snowpark stored procedure though. 
And you're right above about that section of the code in fractribution for shapley being the reason that the rows are left out - i commented out the  `or not path_summary.conversions: ` and there were 6 rows in the report table and three when it was left in.

zloff commented:
> I've re-run it this morning, running with last_touch it seems that the snowpark method returns just the 3 rows, the standard method returns 6. You're right that shapley seems to return 3 for both methods.

once i added the bug fixes for last_touch, i ran snowpark with last_touch and i got 6 rows...

rlh1994 commented:
Okay, removing `or not path_summary.conversions` from the main `.py` file gave me 6 rows for every warehouse, for every model type, using the python one. The snowpark stuff then gave me a load of errors with the tests that I don't have time today to fix, but hopefully you can get them working. I'll also add a comment on the code because I think the tests only test one method still.

I think we need to edit the source data so that at least two of the methods lead to different outputs, currently every model type gives the exact same result so there's no good way to test everything works as expected.

rlh1994 commented on integration_tests/.scripts/integration_test.sh:
```diff
@@ -38,15 +39,29 @@ for db in ${DATABASES[@]}; do
   echo "Snowplow Fractribution integration tests: Execute fractribution models"
 
   eval "dbt run --select snowplow_fractribution --target $db --full-refresh" || exit 1;
```
I think the main run of the package needs to have the model variable as well, otherwise it's always going to run with the default model and then the test views will be based on a different one?

zloff replied:
I changed this - thanks!

zloff commented:
> Okay, removing or not path_summary.conversions from the main .py file gave me 6 rows for every warehouse, for every model type, using the python one. The snowpark stuff then gave me a load of errors with the tests that I don't have time today to fix, but hopefully you can get them working. I'll also add a comment on the code because I think the tests only test one method still.

>I think we need to edit the source data so that at least two of the methods lead to different outputs, currently every model type gives the exact same result so there's no good way to test everything works as expected.

I made these changes but tests are still failing because of the float imprecision thing, not sure how to deal with that

rlh1994 committed:
Squashed commit of the following:

commit 0170fba20bbe25f433255b5b2fdf2a558402b291
Author: Ryan Hill <r.l.hill128@gmail.com>
Date:   Fri Jun 9 12:02:23 2023 +0100

    revert license

commit f990b98bc96ee74ef5709790c9fe5cebaa76c774
Author: zloff <zoe@poplindata.com>
Date:   Fri Jun 9 10:12:35 2023 +1000

    Fix failing tests

commit 1d4ceaedc63197439c6b59e0889b4ebfeeb3fa5e
Merge: 2dad8ff fe197c7
Author: zloff <zoe@poplindata.com>
Date:   Fri Jun 9 08:11:35 2023 +1000

    Resolve merge conflicts

commit fe197c77d54a196393f081230d72cdf0bb6e575a
Author: Ryan Hill <r.l.hill128@gmail.com>
Date:   Thu Jun 8 10:58:13 2023 +0100

    ignore extra column

commit 7e4d5c50aa4dadee5e4c793b70c2be660c7dc203
Author: Ryan Hill <r.l.hill128@gmail.com>
Date:   Thu Jun 8 10:37:22 2023 +0100

    Add equality test for simplicity

commit 91ea5d6d6f5e7988b590a2ba00aa0d8f7295fe5c
Author: zloff <zoe@poplindata.com>
Date:   Thu Jun 8 16:17:39 2023 +1000

    Update integration source data to give different results for last_touch and shapley, make associated changes

commit f54a6c21796500ed0c2e2e7ce03c15151622fe88
Author: zloff <zoe@poplindata.com>
Date:   Tue Jun 6 20:30:36 2023 +1000

    Add bug fixes from redshift pr to snowpark script

commit faae98b78b9ce83f7f275094302063f6b1f5bf1e
Author: zloff <zoe@poplindata.com>
Date:   Mon May 29 15:52:48 2023 +1000

    Remove superfluous comment

commit ccce908380d7d83106e451adda630afd0ace9c11
Author: zloff <zoe@poplindata.com>
Date:   Mon May 29 15:30:16 2023 +1000

    Update integration tests to test two attribution models

commit 7dfdcb4d461e182d6813524e4963cf31aca57cea
Author: Ryan Hill <r.l.hill128@gmail.com>
Date:   Fri May 26 10:28:12 2023 +0100

    Docs + CL

commit 086f6ca9dd14f1d5f9b2ba3a3b453d9be211aebf
Author: zloff <zoe@poplindata.com>
Date:   Mon May 29 09:04:15 2023 +1000

    Add console log and update table for call_snowpark_macros model

commit 296b2c340f9abd425fb71cc6e9489cb777832db4
Author: zloff <zoe@poplindata.com>
Date:   Fri May 26 15:44:10 2023 +1000

    Change licence to community licence

commit fc17552e88c244c4497f49a37956532e0029c979
Author: Ryan Hill <r.l.hill128@gmail.com>
Date:   Thu May 25 13:19:15 2023 +0100

    scope variables, databricks tests fix, simplify enabled logic

commit a9833aed1d1db8f19a3fc2d37676535dc059c069
Author: zloff <zoe@poplindata.com>
Date:   Wed May 24 14:40:00 2023 +1000

    Change quotation marks and enabled clause

commit 623fae7b8e0002f3310849d9e6935e35a9580f2a
Author: zloff <zoe@poplindata.com>
Date:   Wed May 24 13:19:08 2023 +1000

    Move comment out of config

commit af1865684706a038f9a46a434d31e906ce31adda
Author: zloff <zoe@poplindata.com>
Date:   Wed May 24 13:12:48 2023 +1000

    Add snowpark variable to integration dbt_project.yml

commit c46cd6b900cf51c4e26f7ef42a58127e4e7713b5
Author: zloff <zoe@poplindata.com>
Date:   Wed May 24 12:45:28 2023 +1000

    Move stored procedure call to model file, update readme

commit e29e6b373e266489aa9eed3c47057aa3540fccc9
Author: zloff <zoe@poplindata.com>
Date:   Tue May 23 16:17:57 2023 +1000

    Add integration tests for snowpark-created tables

commit ce36666e137ba99a95c1c7c6c0caeaadfa731e14
Author: zloff <zoe@poplindata.com>
Date:   Thu May 18 15:59:36 2023 +1000

    Add if execute blocks and move hooks to model config

commit c658d4cfe9cb017a47441e21396f9f1de6fbf106
Author: zloff <zoe@poplindata.com>
Date:   Wed May 17 11:07:00 2023 +1000

    Add macros to create and run stored procedure on Snowpark

commit 2dad8ff4bbf4dcc686553995b61c9a8a61aa1057
Author: zloff <zoe@poplindata.com>
Date:   Thu Jun 8 16:17:39 2023 +1000

    Update integration source data to give different results for last_touch and shapley, make associated changes

commit 048e3ff66d2c7832974cba64da9a6f6bf360b35e
Author: zloff <zoe@poplindata.com>
Date:   Tue Jun 6 20:30:36 2023 +1000

    Add bug fixes from redshift pr to snowpark script

commit ed04da85fcd635520dc46569d236671e4dfc07cd
Author: zloff <zoe@poplindata.com>
Date:   Mon May 29 15:52:48 2023 +1000

    Remove superfluous comment

commit 2ffbfd14e7e5bf036ef7014cf8b069bc386e102d
Author: zloff <zoe@poplindata.com>
Date:   Mon May 29 15:30:16 2023 +1000

    Update integration tests to test two attribution models

commit 4ded5aafda64ec75161be0c084623c636ce28c69
Merge: 9b3ffa0 62f7521
Author: zloff <zoe@poplindata.com>
Date:   Mon May 29 09:04:32 2023 +1000

    Merge branch 'feature/run_python_on_snowpark' of https://github.com/snowplow/dbt-snowplow-fractribution into feature/run_python_on_snowpark

commit 9b3ffa0f306ccf1589a50b0b87e5c4dfd2573909
Author: zloff <zoe@poplindata.com>
Date:   Mon May 29 09:04:15 2023 +1000

    Add console log and update table for call_snowpark_macros model

commit 62f7521b56a81d75fa73e5674e0d0bca6b26ad37
Author: Ryan Hill <r.l.hill128@gmail.com>
Date:   Fri May 26 10:28:12 2023 +0100

    Docs + CL

commit 262cc622a0e7c301f056a02ae4f11fe94dd0525e
Author: zloff <zoe@poplindata.com>
Date:   Fri May 26 15:44:10 2023 +1000

    Change licence to community licence

commit 7cca2a832b25774f68e22688501cb9f532a461e6
Author: Ryan Hill <r.l.hill128@gmail.com>
Date:   Thu May 25 13:19:15 2023 +0100

    scope variables, databricks tests fix, simplify enabled logic

commit 8aa8bfbf5c97a5af9e31e58f7a0cf03de6ab1d25
Author: zloff <zoe@poplindata.com>
Date:   Wed May 24 14:40:00 2023 +1000

    Change quotation marks and enabled clause

commit 4a3f64c6d0d82065c743a6acd441e056bb3f264e
Author: zloff <zoe@poplindata.com>
Date:   Wed May 24 13:19:08 2023 +1000

    Move comment out of config

commit fb6641ad8fbc5dfcfebfbd44bfa3a8b42be2e2de
Author: zloff <zoe@poplindata.com>
Date:   Wed May 24 13:12:48 2023 +1000

    Add snowpark variable to integration dbt_project.yml

commit 2524c47d59c02021ad2181adfc074f001af5156d
Author: zloff <zoe@poplindata.com>
Date:   Wed May 24 12:45:28 2023 +1000

    Move stored procedure call to model file, update readme

commit a635ac33f64b35aa7c8f0f9a9ba141689e46b027
Author: zloff <zoe@poplindata.com>
Date:   Tue May 23 16:17:57 2023 +1000

    Add integration tests for snowpark-created tables

commit 43ca95e3e3f5cc8e8bd3c3991d30b2cf8e296431
Author: zloff <zoe@poplindata.com>
Date:   Thu May 18 15:59:36 2023 +1000

    Add if execute blocks and move hooks to model config

commit bdfc32b25ad817506aa9d070fa99a6b9e5923744
Author: zloff <zoe@poplindata.com>
Date:   Wed May 17 11:07:00 2023 +1000

    Add macros to create and run stored procedure on Snowpark

agnes-kiss reviewed (approved):
LGTM! I missed the struggle but it looks clean, good job!
